src/shims/unix/socket.rs

```diff
@@ -20,17 +38,122 @@ impl FileDescription for SocketPair {
+
+        if self.is_nonblock {
+            if !readbuf_has_writer {
+                if data_size == 0 {
+                    // Non-blocking socketpair with no writer and empty buffer.
+                    return Ok(Ok(0));
+                }
+            } else if data_size == 0 {
+                // Non-blocking socketpair with writer and empty buffer.
+                // https://linux.die.net/man/2/read
+                // EAGAIN or EWOULDBLOCK can be returned for socket,
+                // POSIX.1-2001 allows either error to be returned for this case.
+                // Since there is no ErrorKind for EAGAIN, WouldBlock is used.
+                return Ok(Err(Error::from(ErrorKind::WouldBlock)));
+            }
+        } else {
+            if !readbuf_has_writer {
+                if data_size == 0 {
+                    // Blocking socketpair with no writer and empty buffer.
+                    return Ok(Ok(0));
+                }
+            } else if data_size == 0 {
+                // Blocking socketpair with writer and empty buffer.
+                // FIXME: blocking is currently not supported
+                throw_unsup_format!("socketpair read: blocking isn't supported yet");
+            }
+        }
+        if let Some(data_race) = &ecx.machine.data_race {
+            data_race.acquire_clock(&self.readbuf.clone().borrow_mut().clock, &ecx.machine.threads);
+        }
+        // Do full read / partial read based on the space available.
+        let actual_read_size = request_byte_size.min(data_size);
+        self.readbuf.borrow_mut().buf.read_exact(&mut bytes[..actual_read_size]).unwrap();
+        return Ok(Ok(actual_read_size));
+    }
+
+    fn write<'tcx>(
+        &mut self,
+        _communicate_allowed: bool,
+        bytes: &[u8],
+        ecx: &mut MiriInterpCx<'tcx>,
+    ) -> InterpResult<'tcx, io::Result<usize>> {
+        let write_size = bytes.len();
+        // Always succeed on write size 0.
+        if write_size == 0 {
+            return Ok(Ok(0));
+        }
+
+        let writebuf = self.writebuf.upgrade();
+        if writebuf.is_none() {
+            // If the upgrade from Weak to Rc fails, it indicates that all read ends have been
+            // closed.
+            return Ok(Err(Error::from(ErrorKind::BrokenPipe)));
+        }
+
+        let data_size = writebuf.clone().unwrap().borrow().buf.len();
+        let available_space = MAX_SOCKETPAIR_BUFFER_CAPACITY.checked_sub(data_size).unwrap();
+        let actual_write_size = write_size.min(available_space);
+        if self.is_nonblock {
+            if write_size <= MAX_SOCKETPAIR_BUFFER_CAPACITY {
+                if available_space < write_size {
+                    // Non-blocking socketpair with write_size <= buffer_size and has
+                    // insufficient space to writer all the data.
+                    return Ok(Err(Error::from(ErrorKind::WouldBlock)));
+                }
+            } else if available_space == 0 {
```

Comment:
Is this reachable? If `available_space == 0`, then the only way not to go into the `if available_space < write_size` is if `write_size == 0`, which we already checked at the start of this function

Reply:
That branch will hit only if:
1. write_size != 0 (checked at the start of the program)
2. socketpair is ``non-blocking``
3. write_size > MAX_SOCKETPAIR_BUFFER_CAPACITY
4. available space == 0

To go into ``if available_space < write_size`` above, it needs to be ``write_size <= MAX_SOCKETPAIR_BUFFER_CAPACITY``; To go into ``available_size <= write_size`` below, it needs to be be ``blocking`` socketpair. 

Let me know if I missed anything. There are a lot of branches and I am not sure if I got them correctly even after multiple checks.

Reply:
Why is `write_size > MAX_SOCKETPAIR_BUFFER_CAPACITY` something useful to check? I would have assumed the `available_space < write_size` check to suffice. Basically the only difference between blocking and nonblocking is the behaviour if we'd block, not how we check whether we block.

Reply:
Specification updated in the latest comment below https://github.com/rust-lang/miri/pull/3609#issuecomment-2142513146